jbpm-workitems/jbpm-workitems-email/src/main/java/org/jbpm/process/workitem/email/TemplateManager.java

```diff
@@ -0,0 +1,201 @@
+        
+        if (watcherEnabled) {
+            this.watcher = new TemplateDirectoryWatcher(templateDirectory);
+            Thread watcherThread = new Thread(watcher, "Email Template Watch Thread");
+            watcherThread.start();
+        }
+    }
+    
+    public static TemplateManager get() {
+        return INSTANCE;
+    }
+    
+    public static synchronized TemplateManager reset() { 
+        // close in case there is watcher running to stop it
+        INSTANCE.close();
+        // create new instance
+        INSTANCE = new TemplateManager();
+        return INSTANCE;
+    }
+    
+    public String render(String templateName, Map<String, Object> parameters) {
+        StringWriter out = new StringWriter();
+        try {
+            Template template = cfg.getTemplate(templateName);
+            
+            template.process(parameters, out);
+        } catch (Exception e) {
+            throw new IllegalArgumentException("Template " + templateName + " not found", e);
+        }
+        return out.toString();
+    }
+    
+    protected void loadTemplates() {
+        if (templateDirectory != null) {
+            File directory = new File(templateDirectory);
+            
+            if (directory.exists() && directory.isDirectory()) {
+                File[] foundTemplates = directory.listFiles((dir, name) -> { return name.endsWith(".html");});
+                
+                for (File templateFile : foundTemplates) {
+                    
+                    loadTemplate(templateFile);
+                }
+            }
+        }
+    }
+    
+    protected void loadTemplate(File templateFile) {
+        String templateId = templateFile.getName().substring(0, templateFile.getName().lastIndexOf("."));
+        
+        try (FileInputStream inputStream = new FileInputStream(templateFile)) {
+            this.stringLoader.putTemplate(templateId, read(inputStream));
+            
+            logger.info("Loaded template {} from file {}", templateId, templateFile);
+        } catch (Exception e) {
+            logger.warn("Exception while loading template from {} due to {}", templateFile, e.getMessage(), e);
+        }
+    }
+    
+    protected void removeTemplate(File templateFile) {
+        String templateId = templateFile.getName().substring(0, templateFile.getName().lastIndexOf("."));
+        stringLoader.removeTemplate(templateId);
+        logger.info("Removed template {} backed by file {}", templateId, templateFile);
+    }
+     
+    protected String read(InputStream input) {
+        String lineSeparator = System.getProperty("line.separator");
+
+        try (BufferedReader buffer = new BufferedReader(new InputStreamReader(input, Charset.forName("UTF-8")))) {
+            return buffer.lines().collect(Collectors.joining(lineSeparator));
+        } catch (Exception e) {
+            return null;
+        }
+    }
+    
+    public void close() {
+        if (this.watcher != null) {
```

Comment:
Looking at this method, is it possible that 2 threads could call it? If the thread A sets `watcher=null` and thread B is just going to call `watcher.stop()`, it will result in NPE. If we omit setting the reference to null, the `stop()` method might get called more than once, but it shouldn't be a problem.

Reply:
done, though in general close should not be called by multiple threads